luna/gateware/usb/usb2/descriptor_single_memory.py

```diff
@@ -0,0 +1,438 @@
+
+    def _test_stall(self, type_number, index, start_position, max_length):
+        """ Triggers a read and checks if correctly stalled. """
+
+        yield self.dut.value.word_select(1, 8).eq(type_number)  # Type
+        yield self.dut.value.word_select(0, 8).eq(index)  # Index
+        yield self.dut.length.eq(max_length)
+        yield self.dut.start_position.eq(start_position)
+        yield self.dut.tx.ready.eq(1)
+        yield self.dut.start.eq(1)
+        yield
+
+        yield self.dut.start.eq(0)
+
+        cycles_passed = 0
+        timeout = 100
+
+        while not (yield self.dut.stall):
+            self.assertEqual((yield self.dut.tx.valid), 0)
+            yield
+
+            cycles_passed += 1
+            if timeout and cycles_passed > timeout:
+                raise RuntimeError(f"Timeout waiting for stall!")
+
+    @usb_domain_test_case
+    def test_all_descriptors(self):
+        for type_number, index, raw_descriptor in self.descriptors:
+            yield from self._test_descriptor(type_number, index, raw_descriptor, 0, len(raw_descriptor))
+            yield from self._test_descriptor(type_number, index, raw_descriptor, 0, len(raw_descriptor), delay_ready=10)
+
+    @usb_domain_test_case
+    def test_all_descriptors_with_offset(self):
+        for type_number, index, raw_descriptor in self.descriptors:
+            if len(raw_descriptor) > 1:
+                yield from self._test_descriptor(type_number, index, raw_descriptor, 1, len(raw_descriptor))
+
+    @usb_domain_test_case
+    def test_all_descriptors_with_length(self):
+        for type_number, index, raw_descriptor in self.descriptors:
+            if len(raw_descriptor) > 1:
+                yield from self._test_descriptor(type_number, index, raw_descriptor, 0, min(8, len(raw_descriptor)-1))
+                yield from self._test_descriptor(type_number, index, raw_descriptor, 0, min(8, len(raw_descriptor)-1), delay_ready=10)
+
+    @usb_domain_test_case
+    def test_all_descriptors_with_offset_and_length(self):
+        for type_number, index, raw_descriptor in self.descriptors:
+            if len(raw_descriptor) > 1:
+                yield from self._test_descriptor(type_number, index, raw_descriptor, 1, min(8, len(raw_descriptor)-1))
+
+    @usb_domain_test_case
+    def test_all_descriptors_with_zero_length(self):
+        for type_number, index, raw_descriptor in self.descriptors:
+            yield from self._test_stall(type_number, index, 0, 0)
+
+    @usb_domain_test_case
+    def test_unavailable_descriptor(self):
+        yield from self._test_stall(StandardDescriptorNumbers.STRING, 100, 0, 64)
+
+    @usb_domain_test_case
+    def test_unavailable_index_type(self):
+        # Unavailable index in between
+        yield from self._test_stall(0x10, 0, 0, 64)
+
+        # Index after last used type
+        yield from self._test_stall(0x42, 0, 0, 64)
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
This is excellent, high quality work. Sorry for nitpicking a bit, but it is a by product of me reading the code. Some parts might be a matter of taste and conditioning, so feel free to do what you like anyway :)